src/libprojectM/Renderer/MilkdropWaveform.cpp

```diff
@@ -4,393 +4,613 @@
-    const float  vol_scale = context.beatDetect->getPCMScale();
-    if (vol_scale != 1.0)
+    //wave color brightening
+    //
+    //forces max color value to 1.0 and scales
+    // the rest accordingly
+    if (mode == MilkdropWaveformMode::Blob2 || mode == MilkdropWaveformMode::ExplosiveHash)
+    {
+        switch (context.texsize)
+        {
+            case 256:
+                temp_a *= 0.07f;
+                break;
+            case 512:
+                temp_a *= 0.09f;
+                break;
+            case 1024:
+                temp_a *= 0.11f;
+                break;
+            case 2048:
+                temp_a *= 0.13f;
+                break;
+        }
+    }
+    else if (mode == MilkdropWaveformMode::Blob3)
     {
-        for (int i=0 ; i<512 ; i++)
+        switch (context.texsize)
         {
-            pcmdataL[i] *= vol_scale;
-            pcmdataR[i] *= vol_scale;
+            case 256:
+                temp_a *= 0.075f;
+                break;
+            case 512:
+                temp_a *= 0.15f;
+                break;
+            case 1024:
+                temp_a *= 0.22f;
+                break;
+            case 2048:
+                temp_a *= 0.33f;
+                break;
         }
+        temp_a *= 1.3f;
+        temp_a *= std::pow(context.beatDetect->treb, 2.0f);
     }
 
-    float r2, theta;
+    if (temp_a < 0.0f)
+    {
+        temp_a = 0.0f;
+    }
 
-	float wave_x_temp=0;
-	float wave_y_temp=0;
+    if (temp_a > 1.0f)
+    {
+        temp_a = 1.0f;
+    }
 
-	float cos_rot;
-	float sin_rot;
+    if (maximizeColors)
+    {
+        constexpr float fMaximizeWaveColorAmount = 1.0f;
+        float cr{ r };
+        float cg{ g };
+        float cb{ b };
+
+        float max = cr;
+        if (max < cg) max = cg;
+        if (max < cb) max = cb;
+        if (max > 0.01f)
+        {
+            cr = cr / max * fMaximizeWaveColorAmount + cr * (1.0f - fMaximizeWaveColorAmount);
+            cg = cg / max * fMaximizeWaveColorAmount + cg * (1.0f - fMaximizeWaveColorAmount);
+            cb = cb / max * fMaximizeWaveColorAmount + cb * (1.0f - fMaximizeWaveColorAmount);
+        }
 
-    const float temp_y = -1*(y-1.0f);
+        glVertexAttrib4f(1, cr, cg, cb, temp_a * masterAlpha);
+    }
+    else
+    {
+        glVertexAttrib4f(1, r, g, b, temp_a * masterAlpha);
+    }
+}
 
-	two_waves = false;
-	loop = false;
 
-	switch(mode)
-	{
+void MilkdropWaveform::WaveformMath(RenderContext& context)
+{
+    constexpr float PI{ 3.14159274101257324219f };
 
-		case Circle:
-	  	{
- 		    loop = true;
-			rot =   0;
-			aspectScale=1.0;
+    delete[] wave1Vertices;
+    wave1Vertices = nullptr;
 
-            samples = 512;
+    delete[] wave2Vertices;
+    wave2Vertices = nullptr;
 
-			float inv_nverts_minus_one = 1.0f/(float)(samples);
+    // NOTE: Buffer size is always 512 samples, waveform points 480 or less since some waveforms use a positive offset!
+    float pcmDataL[512]{ 0.0f };
```

Comment:
Can we use a constant for the 512 please?

Reply:
Sure thing will add a constexpr at the top of the function and also turn this into `std::array<float,n>`.